v2/yarpcgrpc/dialer.go

```diff
@@ -21,30 +21,82 @@
 package grpc
 
 import (
-	"go.uber.org/yarpc/api/peer"
+	"context"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/yarpcpeer"
 )
 
-// NewDialer creates a transport that is decorated to retain peers with
-// additional gRPC dial options.
-func (t *Transport) NewDialer(options ...DialOption) *Dialer {
-	return &Dialer{trans: t, options: newDialOptions(options)}
-}
+var _ yarpc.Dialer = (*Dialer)(nil)
 
 // Dialer is a decorator for a gRPC transport that threads dial options for
 // every retained peer.
 type Dialer struct {
-	trans   *Transport
-	options *dialOptions
+	lock          sync.Mutex
+	options       *dialOptions
+	addressToPeer map[string]*grpcPeer
 }
 
-var _ peer.Transport = (*Dialer)(nil)
+// Start starts the gRPC dialer.
+func (d *Dialer) Start(_ context.Context) error {
+	d.addressToPeer = make(map[string]*grpcPeer)
+	// init options?
+	return nil
+}
+
+// Stop stops the gRPC dialer.
+func (d *Dialer) Stop() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	for _, grpcPeer := range d.addressToPeer {
+		grpcPeer.stop()
+	}
+	var err error
+	for _, grpcPeer := range d.addressToPeer {
+		err = multierr.Append(err, grpcPeer.wait())
+	}
+	return err
+}
 
 // RetainPeer retains the identified peer, passing dial options.
-func (d *Dialer) RetainPeer(id peer.Identifier, ps peer.Subscriber) (peer.Peer, error) {
-	return d.trans.retainPeer(id, d.options, ps)
+func (d *Dialer) RetainPeer(id yarpc.Identifier, ps yarpc.Subscriber) (yarpc.Peer, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
```

Comment:
This might not be an issue for the time being, but we could use finer-grained locking here, i.e. using a read lock to retrieve the peer, and a write lock only when it is being added to our map. This would let us remove the cost of `defer`, too.

Reply:
note: we use different locks for the internals, I think I'd rather leave this as is to keep each operation one atomic event.